File: 83. Remove Duplicates from Sorted List.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        visit=head
        s=head
        while visit!= None:
            #we only need to connect with the different value nodes.
            if s.val != visit.val:
                s.next=visit
                s=visit
            #if we visit the last node but it's value is equal as s:
            elif visit.next ==None and s.val==visit.val:
                s.next=None #which means we don't need that link.
            #Keep doing forward visit.
            visit = visit.next
        return head

File: 83. Remove Duplicates from Sorted List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        if (!head){
            return nullptr;
        }
        ListNode* prev = head;
        int dul_val = head->val;
        ListNode* curr = head->next;
        while (curr){
            if (curr->val != dul_val){
                dul_val = curr->val;
                prev->next = curr;
                prev = curr;
            }
            curr = curr->next;
        }
        if (prev->val == dul_val){
            prev->next = nullptr;
        }
        return head; 
    }
};
